src/mlpack/methods/neighbor_search/neighbor_search.hpp

```diff
@@ -309,14 +412,21 @@ class NeighborSearch
   //! Return the number of node combination scores during the last search.
   size_t Scores() const { return scores; }
 
+  //! Access the search mode.
+  NeighborSearchMode SearchMode() const { return searchMode; }
+  //! Modify the search mode.
+  void SetSearchMode(const NeighborSearchMode mode);
```

Comment:
I think I'd prefer this to be `NeighborSearchMode& SearchMode()` instead of the way it is now, so it can operate the same way the other functions did.  I realize that this can cause some difficulty (this is similar to why you need `UpdateSearchMode()`), but I think we can resolve any actual problems from this by saying that calling `SingleMode()` or `Naive()` (until mlpack 3.0.0 when they are gone) will override whatever `SearchMode()` is set to.

Alternately, we remove `SearchMode()` and `SetSearchMode()` and add a `greedy` member and do the same thing that you are doing with `naive` and `singleMode`.  But I like that solution less.